llama.cpp

```diff
@@ -7269,13 +7483,18 @@ static int llama_decode_internal(
 
     // the output is always the last tensor in the graph
     struct ggml_tensor * res = gf->nodes[gf->n_nodes - 1];
-    GGML_ASSERT(strcmp(res->name, "result_output") == 0);
-
-    // the embeddings could be the second to last tensor, or the third to last tensor
-    struct ggml_tensor * embeddings = gf->nodes[gf->n_nodes - 2];
-    if (strcmp(embeddings->name, "result_norm") != 0) {
-        embeddings = gf->nodes[gf->n_nodes - 3];
-        GGML_ASSERT(strcmp(embeddings->name, "result_norm") == 0);
+    struct ggml_tensor * embeddings = nullptr;
+    if (strcmp(res->name, "result_embed") == 0) {
+        embeddings = res;
+        res = nullptr;
+    } else {
+        // the embeddings could be the second to last tensor, or the third to last tensor
+        GGML_ASSERT(strcmp(res->name, "result_output") == 0);
+        embeddings = gf->nodes[gf->n_nodes - 2];
+        if (strcmp(embeddings->name, "result_norm") != 0) {
+            embeddings = gf->nodes[gf->n_nodes - 3];
+            GGML_ASSERT(strcmp(embeddings->name, "result_norm") == 0);
+        }
```

Comment:
We could probably replace this with `ggml_graph_get_tensor(gf, "result_norm")` instead of guessing where it is.

Reply:
Simplified the logic a bit using the `ggml_graph_get_tensor` approach. I think it makes sense?